### ISSUE TITLE
Enable Ragged Rollouts: Allow Variable-Length list[sample] per Batch

### ISSUE THREAD

**Issue by adol001** [THUDM/slime#378]
Could we support padding-style samples—i.e., samples that are masked out so they don’t participate in weight updates or advantage computation? The main use case is when each rollout returns a variable number of samples. My current approach is to pad with dummy samples to force every rollout to return a fixed-length list[sample]. 

Even better would be supporting variable-length list[sample] within a batch. For example, with rollout_batch_size = 1 and n_samples_per_prompt = 2, the first rollout might return one sample and the second rollout two samples—giving 3 in total, which can’t be evenly divided by 1 * 2

This happens with longer conversations, where one sample has to be split into multiple samples. Consequently, the length of the list[sample] returned by a single rollout becomes variable.

**Comment by zhuzilin** [comment]
maybe the multi-agent example could help.

**Comment by adol001** [comment]
> maybe the multi-agent example could help.

@zhuzilin 

The current rollout method supports returning a variable-length sample list, but the present logic is as follows:

```
args.global_batch_size = args.rollout_batch_size * args.n_samples_per_prompt // args.num_steps_per_rollout

if len(data) % self.args.global_batch_size != 0:
    trim_len = (len(data) // self.args.global_batch_size) * self.args.global_batch_size
    origin_data_length = len(data)
    data = data[:trim_len]
    print(f"trim number of samples from {origin_data_length} to {trim_len}")
```

If the rollout batch size is 1 and args.n_samples_per_prompt is 2, and the first rollout returns 1 sample while the second rollout returns 2 samples, then the total 3 is not divisible by 2. According to the current logic, the data will be trimmed.

To fully support rollouts that return variable-length sample lists, you need to make sure downstream steps—such as advantage (ADV) computation and Megatron loss calculation—explicitly handle variable training set sizes. Alternatively, you can use a slightly tricky approach: pad samples so that each rollout returns an aligned number of samples, apply a loss mask of 0 to the padded entries, and ensure this does not affect ADV computation for the real data.
